### PR TITLE
netty: fix static data race on handler settings

### PR DESCRIPTION
**Pull request by carl-mastrangelo**
As seen here:

```
  Write of size 8 at 0x7f0046595478 by thread T101 (mutexes: write M590111189888139032):
    #0 io.grpc.netty.HandlerSettings.setAutoWindow(Lio/grpc/netty/AbstractNettyHandler;)V (HandlerSettings.java:53)  
    #1 io.grpc.netty.NettyClientTransport.start(Lio/grpc/internal/ManagedClientTransport$Listener;)Ljava/lang/Runnable; (NettyClientTransport.java:148)  
    #2 io.grpc.internal.ForwardingConnectionClientTransport.start(Lio/grpc/internal/ManagedClientTransport$Listener;)Ljava/lang/Runnable; (ForwardingConnectionClientTransport.java:45)  
    #3 io.grpc.internal.TransportSet.startNewTransport(Lio/grpc/internal/DelayedClientTransport;)Ljava/lang/Runnable; (TransportSet.java:234)  
    #4 io.grpc.internal.TransportSet.obtainActiveTransport()Lio/grpc/internal/ClientTransport; (TransportSet.java:204)  
    #5 io.grpc.internal.ManagedChannelImpl$3.getTransport(Lio/grpc/EquivalentAddressGroup;)Lio/grpc/internal/ClientTransport; (ManagedChannelImpl.java:721)  
    #6 io.grpc.internal.ManagedChannelImpl$3.getTransport(Lio/grpc/EquivalentAddressGroup;)Ljava/lang/Object; (ManagedChannelImpl.java:660)  
    #7 io.grpc.PickFirstBalancerFactory$PickFirstBalancer$1.get()Ljava/lang/Object; (PickFirstBalancerFactory.java:132)  
    #8 io.grpc.internal.DelayedClientTransport$2.run()V (DelayedClientTransport.java:264)  

  Previous write of size 8 at 0x7f0046595478 by thread T94 (mutexes: write M576881865974269888):
    #0 io.grpc.netty.HandlerSettings.setAutoWindow(Lio/grpc/netty/AbstractNettyHandler;)V (HandlerSettings.java:53)  
    #1 io.grpc.netty.NettyClientTransport.start(Lio/grpc/internal/ManagedClientTransport$Listener;)Ljava/lang/Runnable; (NettyClientTransport.java:148)  
    #2 io.grpc.internal.ForwardingConnectionClientTransport.start(Lio/grpc/internal/ManagedClientTransport$Listener;)Ljava/lang/Runnable; (ForwardingConnectionClientTransport.java:45)  
    #3 io.grpc.internal.TransportSet.startNewTransport(Lio/grpc/internal/DelayedClientTransport;)Ljava/lang/Runnable; (TransportSet.java:234)  
    #4 io.grpc.internal.TransportSet.obtainActiveTransport()Lio/grpc/internal/ClientTransport; (TransportSet.java:204)  
    #5 io.grpc.internal.ManagedChannelImpl$3.getTransport(Lio/grpc/EquivalentAddressGroup;)Lio/grpc/internal/ClientTransport; (ManagedChannelImpl.java:721)  
    #6 io.grpc.internal.ManagedChannelImpl$3.getTransport(Lio/grpc/EquivalentAddressGroup;)Ljava/lang/Object; (ManagedChannelImpl.java:660)  
    #7 io.grpc.PickFirstBalancerFactory$PickFirstBalancer$1.get()Ljava/lang/Object; (PickFirstBalancerFactory.java:132)  
    #8 io.grpc.internal.DelayedClientTransport$2.run()V (DelayedClientTransport.java:264)  

```

### CODE REVIEW

File: netty/src/main/java/io/grpc/netty/HandlerSettings.java [hunk 1]
```diff
@@ -41,36 +41,47 @@
...
+      } else if (handler instanceof NettyServerHandler) {
+        serverHandler = handler;
+      } else {
+        throw new RuntimeException("Expecting NettyClientHandler or NettyServerHandler");
+      }
     }
   }
 
-  public static void autoWindowOn(boolean autoFlowControl) {
+  public static void enable(boolean enable) {
```
Comment (ejona86) [inline netty/src/main/java/io/grpc/netty/HandlerSettings.java]: What is the point of `enable`? It seems like the synchronization is enough by itself.

Comment (carl-mastrangelo) [inline netty/src/main/java/io/grpc/netty/HandlerSettings.java]: To make it a noop when not running in a test.  It also makes it skip synchronization.

**Comment by carl-mastrangelo** [comment]
As an aside, this github review system is getting to be annoying.  Rebasing to master forces a very slow travis and jenkins build, and blocks submission.  

I don't mind running the tests, but it makes submission take a lot longer since I can't auto submit on the tests passing.  Now I have to manually poll this page to see when I can submit.

**Comment by carl-mastrangelo** [comment]
retest this please

**Comment by ejona86** [comment]
> Rebasing to master

With the GitHub review, you don't need to push out a rebase first. Instead, you can just click the "Rebase and merge" button that avoids the retesting wait.

If you rebase first manually, then pushing manually seems fine.

**Comment by carl-mastrangelo** [comment]
You can't actually push manually, github prevents you:

```
Wed Oct 26 01:55:11$ git push upstream
Counting objects: 20, done.
Delta compression using up to 12 threads.
Compressing objects: 100% (13/13), done.
Writing objects: 100% (20/20), 1.50 KiB | 0 bytes/s, done.
Total 20 (delta 7), reused 0 (delta 0)
remote: Resolving deltas: 100% (7/7), completed with 7 local objects.
remote: error: GH006: Protected branch update failed for refs/heads/master.
remote: error: 2 of 3 required status checks have not succeeded: 1 failing and 1 pending.
To git@github.com:grpc/grpc-java.git
 ! [remote rejected] master -> master (protected branch hook declined)
error: failed to push some refs to 'git@github.com:grpc/grpc-java.git'
~/git/grpc-java
```

**Comment by ejona86** [comment]
CC @nicolasnoble 

> remote: error: 2 of 3 required status checks have not succeeded: 1 failing and 1 pending.

Whoa. Reading some of the Settings, I feel like it may have changed behavior, or at least documentation changed.

I've changed it so Travis-CI is no longer Required, since I agree OS X Travis is consistently taking forever. I'm not sure what we want long-term.

**Pull request merged**